File: algorithms/spfa.py
from collections import deque
from typing import List, Optional, Tuple, Any
import networkx as nx


def spfa(
    G: nx.Graph,
    source: Any,
    target: Any
) -> Optional[Tuple[List[Any], float, List[Any]]]:
    """
    SPFA (Shortest Path Faster Algorithm)
    - Cải tiến của Bellman-Ford, dùng queue để tối ưu hóa.
    - Hoạt động tốt với đồ thị có cạnh âm (nhưng không có chu trình âm).

    Parameters
    ----------
    G : nx.Graph
        Đồ thị có trọng số (có thể âm).
    source : Any
        Node bắt đầu.
    target : Any
        Node đích.

    Returns
    -------
    (path, cost, visited_order)
        path : danh sách node trên đường đi ngắn nhất.
        cost : tổng chi phí đường đi.
        visited_order : thứ tự các node được lấy ra khỏi queue (duyệt relax).
    """

    dist = {node: float('inf') for node in G.nodes}
    parent = {node: None for node in G.nodes}
    in_queue = {node: False for node in G.nodes}
    count = {node: 0 for node in G.nodes}  # số lần node được relax
    dist[source] = 0

    q = deque([source])
    in_queue[source] = True
    visited_order = []

    while q:
        u = q.popleft()
        in_queue[u] = False
        visited_order.append(u)

        for v in G.neighbors(u):
            w = G[u][v].get('weight', 1)
            if dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                parent[v] = u
                if not in_queue[v]:
                    q.append(v)
                    in_queue[v] = True
                    count[v] += 1
                    # Phát hiện chu trình âm (nếu node được relax quá nhiều lần)
                    if count[v] > len(G.nodes):
                        raise ValueError("Graph contains a negative-weight cycle")

    # Nếu không tìm được đường đi
    if dist[target] == float('inf'):
        return None

    # Truy vết đường đi ngắn nhất
    path = []
    curr = target
    while curr is not None:
        path.append(curr)
        curr = parent[curr]
    path.reverse()

    return path, dist[target], visited_order


if __name__ == "__main__":
    G = nx.DiGraph()
    edges = [
        ('A', 'B', 2),
        ('A', 'C', 5),
        ('B', 'C', 3),
        ('B', 'D', 4),
        ('C', 'D', 2)
    ]
    G.add_weighted_edges_from(edges)

    result = spfa(G, 'A', 'D')
    print("Path:", result[0])
    print("Cost:", result[1])
    print("Visited order:", result[2])

File: webapp/src/algorithms/spfa.js
// SPFA (Shortest Path Faster Algorithm) with step-by-step visualization

import { getNeighbors } from '../utils/graphUtils';

/**
 * SPFA algorithm implementation with event emission for visualization
 * An optimization of Bellman-Ford using a queue
 * @param {Object} graph - Graph object with nodes and edges
 * @param {string} source - Source node id
 * @param {string} target - Target node id
 * @param {boolean} isDirected - Whether the graph is directed
 * @returns {Object} - Result with path, cost, steps, and visited order
 */
export function spfa(graph, source, target, isDirected = false) {
  const steps = [];
  const visitOrder = [];
  const dist = new Map();
  const parent = new Map();
  const inQueue = new Map();
  const count = new Map();
  const queue = [source];
  const n = graph.nodes.size;

  // Initialize
  for (const [id] of graph.nodes) {
    dist.set(id, Infinity);
    parent.set(id, null);
    inQueue.set(id, false);
    count.set(id, 0);
  }

  dist.set(source, 0);
  inQueue.set(source, true);

  // Initial step
  steps.push({
    type: 'init',
    current: null,
    queue: [...queue],
    visited: [],
    frontier: [...queue],
    parent: Object.fromEntries(parent),
    distances: Object.fromEntries(dist),
    message: `Starting SPFA from node ${source}`,
  });

  while (queue.length > 0) {
    const current = queue.shift();
    inQueue.set(current, false);
    visitOrder.push(current);

    // Expand step
    steps.push({
      type: 'expand',
      current,
      queue: [...queue],
      visited: [...visitOrder],
      frontier: [...queue],
      parent: Object.fromEntries(parent),
      distances: Object.fromEntries(dist),
      message: `Processing node ${current} with distance ${dist.get(current).toFixed(1)}`,
    });

    const neighbors = getNeighbors(graph, current, isDirected);

    for (const { node: neighbor, weight } of neighbors) {
      const newDist = dist.get(current) + weight;

      if (newDist < dist.get(neighbor)) {
        const oldDist = dist.get(neighbor);
        dist.set(neighbor, newDist);
        parent.set(neighbor, current);

        steps.push({
          type: 'frontier',
          current,
          neighbor,
          queue: [...queue],
          visited: [...visitOrder],
          frontier: [...queue],
          parent: Object.fromEntries(parent),
          distances: Object.fromEntries(dist),
          oldDist,
          newDist,
          message: `Relaxed ${current} → ${neighbor}: ${oldDist === Infinity ? '∞' : oldDist.toFixed(1)} → ${newDist.toFixed(1)}`,
        });

        if (!inQueue.get(neighbor)) {
          queue.push(neighbor);
          inQueue.set(neighbor, true);
          count.set(neighbor, count.get(neighbor) + 1);

          // Check for negative cycle
          if (count.get(neighbor) > n) {
            steps.push({
              type: 'negative_cycle',
              current,
              neighbor,
              visited: [...visitOrder],
              message: `Negative cycle detected at node ${neighbor}!`,
            });

            return {
              found: false,
              path: [],
              cost: -Infinity,
              steps,
              visitOrder,
              nodesVisited: visitOrder.length,
              negativeCycle: true,
            };
          }
        }
      }
    }
  }

  // Check if target is reachable
  if (dist.get(target) === Infinity) {
    steps.push({
      type: 'no_path',
      visited: [...visitOrder],
      distances: Object.fromEntries(dist),
      message: `No path found from ${source} to ${target}`,
    });

    return {
      found: false,
      path: [],
      cost: Infinity,
      steps,
      visitOrder,
      nodesVisited: visitOrder.length,
    };
  }

  // Reconstruct path
  const path = reconstructPath(parent, target);

  steps.push({
    type: 'goal',
    current: target,
    path,
    cost: dist.get(target),
    visited: [...visitOrder],
    distances: Object.fromEntries(dist),
    message: `Found path: ${path.join(' → ')} with cost ${dist.get(target).toFixed(1)}`,
  });

  return {
    found: true,
    path,
    cost: dist.get(target),
    steps,
    visitOrder,
    nodesVisited: visitOrder.length,
  };
}

/**
 * Reconstruct path from parent map
 */
function reconstructPath(parent, target) {
  const path = [];
  let current = target;
  while (current !== null) {
    path.unshift(current);
    current = parent.get(current);
  }
  return path;
}

export default spfa;
